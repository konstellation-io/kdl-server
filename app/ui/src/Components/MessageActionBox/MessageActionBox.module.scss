@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.container {
  margin: 3 * $grid-unit;
  padding: 3 * $grid-unit;

  background-color: palette(base, 500);
  border: solid 1px palette(lowlight, 900);
  border-radius: 0.5 * $grid-unit;

  &.warn {
    border-color: palette(alert, 700);
  }
  &.error {
    border-color: palette(error, 700);
  }

  .title {
    @include font-small;
    color: font-color(light);

    margin: 0 0 2 * $grid-unit 0;
  }

  .description {
    @include font-small;
    color: font-color(regular);

    margin: 0;
  }
}

.action {
  display: flex;
  align-items: center;
  width: 100%;
  justify-content: flex-end;
  margin-top: 3 * $grid-unit;

  .confirmation {
    display: flex;
    align-items: center;

    .confirmationText {
      @include font-small;
      color: font-color(lowlight);

      margin: 0 4 * $grid-unit 0 2 * $grid-unit;
    }
  }
}

.button {
  width: 20 * $grid-unit;
}
