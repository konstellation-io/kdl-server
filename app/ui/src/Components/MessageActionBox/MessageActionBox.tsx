import { BUTTON_THEMES, Button, Check } from 'kwc';
import React, { FunctionComponent, useEffect, useState } from 'react';

import { OverrideProps } from '@material-ui/core/OverridableComponent';
import { SvgIconTypeMap } from '@material-ui/core';
import cx from 'classnames';
import styles from './MessageActionBox.module.scss';

export enum BOX_THEME {
  DEFAULT = 'default',
  WARN = 'warn',
  ERROR = 'error',
}

const toButtonTheme = new Map([
  [BOX_THEME.DEFAULT, BUTTON_THEMES.DEFAULT],
  [BOX_THEME.WARN, BUTTON_THEMES.WARN],
  [BOX_THEME.ERROR, BUTTON_THEMES.ERROR],
]);

type Action = {
  needConfirmation?: boolean;
  message?: string;
  label: string;
  onClick: () => void;
  Icon?: FunctionComponent<OverrideProps<SvgIconTypeMap<{}, 'svg'>, 'svg'>>;
};

type Props = {
  title: string;
  desciption: string;
  action?: Action;
  theme?: BOX_THEME;
  loading?: boolean;
};
function MessageActionBox({
  title,
  desciption,
  action,
  theme = BOX_THEME.DEFAULT,
  loading = false,
}: Props) {
  const [confirmed, setConfirmed] = useState(false);

  useEffect(() => {
    if (loading) {
      setConfirmed(false);
    }
  }, [loading]);

  return (
    <div
      className={cx(styles.container, styles[theme])}
      onClick={(e) => e.stopPropagation()}
    >
      <p className={styles.title}>{title}</p>
      <p className={styles.description}>{desciption}</p>
      {action && (
        <div className={styles.action}>
          {action.needConfirmation && (
            <div className={styles.confirmation}>
              <Check checked={confirmed} onChange={(v) => setConfirmed(v)} />
              <p className={styles.confirmationText}>{action.message}</p>
            </div>
          )}
          <div className={styles.button}>
            <Button
              label={action.label}
              theme={toButtonTheme.get(theme)}
              onClick={action.onClick}
              disabled={action.needConfirmation && !confirmed}
              loading={loading}
              height={30}
              Icon={action.Icon}
              primary
            />
          </div>
        </div>
      )}
    </div>
  );
}

export default MessageActionBox;
