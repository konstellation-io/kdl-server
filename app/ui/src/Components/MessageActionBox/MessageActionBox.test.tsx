import { Button, Check } from 'kwc';
import MessageActionBox, { BOX_THEME } from './MessageActionBox';

import React from 'react';
import SomeIcon from '@material-ui/icons/Archive';
import { shallow } from 'enzyme';

const onActionMock = jest.fn();

const component = shallow(
  <MessageActionBox
    title="Some title"
    description="Some description"
    action={{
      needConfirmation: true,
      message: 'Some message',
      label: 'Some label',
      onClick: onActionMock,
      Icon: SomeIcon,
    }}
    theme={BOX_THEME.DEFAULT}
  />
);

describe('MessageActionBox component', () => {
  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right texts', () => {
    // Arrange.
    // Act.
    // Assert.
    expect(component.contains('Some title')).toBeTruthy();
    expect(component.contains('Some description')).toBeTruthy();
    expect(component.contains('Some message')).toBeTruthy();
  });

  it('checks validation before performing action', () => {
    // Arrange.
    const buttonInitialDisabled = component.find(Button).props().disabled;

    // Act.
    component.find(Check).props().onChange(true);
    component.find(Button).props().onClick();

    // Assert.
    expect(buttonInitialDisabled).toBeTruthy();
    expect(component.find(Button).props().disabled).toBeFalsy();
    expect(onActionMock).toHaveBeenCalledTimes(1);
  });
});
