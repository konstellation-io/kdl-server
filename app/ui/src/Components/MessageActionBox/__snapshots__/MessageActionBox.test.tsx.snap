// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageActionBox component Component match snapshot 1`] = `
<div
  className="container default"
  onClick={[Function]}
>
  <p
    className="title"
  >
    Some title
  </p>
  <p
    className="description"
  >
    Some description
  </p>
  <div
    className="action"
  >
    <div
      className="confirmation"
    >
      <J
        checked={false}
        onChange={[Function]}
      />
      <p
        className="confirmationText"
      >
        Some message
      </p>
    </div>
    <div
      className="button"
    >
      <T
        Icon={
          Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        disabled={true}
        height={30}
        label="Some label"
        onClick={[MockFunction]}
        primary={true}
        theme="default"
      />
    </div>
  </div>
</div>
`;
